framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -517,18 +544,40 @@ protected boolean visitTypeArgs(
             : AnnotationUtils.getElementValueArray(
                 covariantAnno, covariantValueElement, Integer.class);
 
-    for (int i = 0; i < supertypeTypeArgs.size(); i++) {
-      final AnnotatedTypeMirror superTypeArg = supertypeTypeArgs.get(i);
-      final AnnotatedTypeMirror subTypeArg = subtypeTypeArgs.get(i);
-      final boolean covariant = covariantArgIndexes != null && covariantArgIndexes.contains(i);
-
-      boolean result = isContainedBy(subTypeArg, superTypeArg, covariant);
+    // JLS: 4.10.2. Subtyping among Class and Interface Types
+    // 4th paragraph
+    if (isContainedMany(subtype.getTypeArguments(), supertypeTypeArgs, covariantArgIndexes)) {
+      return true;
+    }
+    // 5th paragraph
+    AnnotatedDeclaredType capturedSubtype =
+        (AnnotatedDeclaredType) subtype.atypeFactory.applyCaptureConversion(subtype);
+    return isContainedMany(
```

Comment:
I am a bit confused by this code.  The 5th paragraph says that -- after capture conversion -- one should do a regular subtyping test.  Is this an optimization, because the subtyping test would boil down to `isContainedMany`?  Or is there another reason we need to diverge from the JLS text?

Reply:
Here's that section in the JLS.  I've numbered the paragraphs and used letters for the bullet points.
> 4.) Given a generic type declaration C<F1,...,Fn> (n > 0), the direct supertypes of the parameterized type C<T1,...,Tn>, where Ti (1 ≤ i ≤ n) is a type, are all of the following:
> 
> A.) D<U1 θ,...,Uk θ>, where D<U1,...,Uk> is a generic type which is a direct supertype of the generic type C<F1,...,Fn> and θ is the substitution [F1:=T1,...,Fn:=Tn].
> 
> B.) C<S1,...,Sn>, where Si contains Ti (1 ≤ i ≤ n) (§4.5.1).
> 
> C.) The type Object, if C<F1,...,Fn> is a generic interface type with no direct superinterfaces.
> 
> D.) The raw type C.
> 
> 5.) Given a generic type declaration C<F1,...,Fn> (n > 0), the direct supertypes of the parameterized type C<R1,...,Rn> where at least one of the Ri (1 ≤ i ≤ n) is a wildcard type argument, are the direct supertypes of the parameterized type C<X1,...,Xn> which is the result of applying capture conversion to C<R1,...,Rn> (§5.1.10).

4A is applied before `visitTypeArgs` is called,  this happens in `visitDeclared_Declared`, `AnnotatedTypes.castedAsSuper(subtype.atypeFactory, subtype, supertype);`.   So subtype and supertype have the same underlying type in `visitTypeArgs`.  (The documentation of subtype and supertype was wrong; I've fixed it.)

4B is applied in `visitTypeArgs` in the if statement `isContainedMany(subtype.getTypeArguments(), supertypeTypeArgs, covariantArgIndexes)` (In your [comment above](https://github.com/typetools/checker-framework/pull/3922/files/8d2e1d98be7e04c8cae2d2da14bc6029af7ab759#r661633409) you asked if this if statement was for 4C, but it's 4B.)

5 is applied in `visitTypeArgs` in `isContainedMany(capturedSubtype.getTypeArguments(), supertypeTypeArgs, covariantArgIndexes)`.  As you suggest, this is an optimization, because calling isSubtype(subtype, supertype) would result in  the `isContainedMany(capturedSubtype.getTypeArguments(), supertypeTypeArgs, covariantArgIndexes)` call.

I've attempted to make this clearer in the code.